File: Reviewmerter/src/main/java/com/reviewmeter/tfg/service/resenaService.java
package com.reviewmeter.tfg.service;

import java.time.LocalDate;
import java.util.List;


import com.reviewmeter.tfg.model.Resena;

public interface resenaService {

	List<Resena> getResenas();

	Resena getResena(Long id);

	Resena crearResena(Resena resena);

	Resena actualizarResena(Long id, Resena resenaActualizada);

	String borrarResena(Long id);

	List<Resena> filtrarPorFechas(LocalDate inicio, LocalDate fin);

	List<Resena> filtrarPorPuntuacion(int min);

	List<Resena> ordenarPorPuntuacionDesc();

	List<Resena> ordenarPorPuntuacionAsc();

	List<Resena> getResenasPorProducto(Long idProducto);

	List<Resena> getResenasPorProductoOrdenDesc(Long idProducto);

	List<Resena> getResenasPorProductoOrdenAsc(Long idProducto);

	List<Resena> getResenasPorProductoYPuntuacion(Long idProducto, int min);

	List<Resena> getResenasPorUsuario(Long idUsuario);
}

File: Tfg-frontend/src/app/Services/resena-service.ts
import { Injectable } from '@angular/core';
import { Resena } from '../Model/resenaModel.Model';
import { Observable } from 'rxjs';
import { HttpClient, HttpParams } from '@angular/common/http';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root',
})
export class ResenaService {

  private baseUrl = `${environment.apiUrl}/api/resenas`;

  constructor(private http: HttpClient) { }

  // GET: Todas las reseñas
  getResenas(): Observable<Resena[]> {
    return this.http.get<Resena[]>(this.baseUrl);
  }

  // GET: Una reseña por ID
  getResena(id: number): Observable<Resena> {
    return this.http.get<Resena>(`${this.baseUrl}/${id}`);
  }

  crearResena(resena: Resena): Observable<Resena> {
    const token = localStorage.getItem('token');
    const headers = { Authorization: `Bearer ${token}` };
    return this.http.post<Resena>(this.baseUrl, resena, { headers });
  }

  actualizarResena(id: number, resena: Resena): Observable<Resena> {
    const token = localStorage.getItem('token');
    const headers = { Authorization: `Bearer ${token}` };
    return this.http.put<Resena>(`${this.baseUrl}/${id}`, resena, { headers });
  }

  borrarResena(id: number): Observable<string> {
    const token = localStorage.getItem('token');
    const headers = { Authorization: `Bearer ${token}` };
    return this.http.delete(`${this.baseUrl}/${id}`, { headers, responseType: 'text' });
  }

  filtrarPorFechas(inicio: string, fin: string): Observable<Resena[]> {
    const params = new HttpParams()
      .set('inicio', inicio)
      .set('fin', fin);
    return this.http.get<Resena[]>(`${this.baseUrl}/filtro-fechas`, { params });
  }

  filtrarPorPuntuacion(min: number): Observable<Resena[]> {
    return this.http.get<Resena[]>(`${this.baseUrl}/puntuacion/${min}`);
  }

  ordenarPorPuntuacionDesc(): Observable<Resena[]> {
    return this.http.get<Resena[]>(`${this.baseUrl}/orden-desc`);
  }

  ordenarPorPuntuacionAsc(): Observable<Resena[]> {
    return this.http.get<Resena[]>(`${this.baseUrl}/orden-asc`);
  }

  // Acepta orden y puntuación mínima como filtros opcionales
  getResenasPorProducto(idProducto: number, orden?: string, minPuntuacion?: number): Observable<any[]> {
    let params: any = {};
    if (orden) params.orden = orden;
    if (minPuntuacion && minPuntuacion > 0) params.minPuntuacion = minPuntuacion.toString();
    
    return this.http.get<any[]>(`${this.baseUrl}/producto/${idProducto}`, { params });
  }

  // Devuelve las 6 reseñas con mejor puntuación para mostrar en el home
  getResenasDestacadas(): Observable<any[]> {
    return this.http.get<any[]>(`${this.baseUrl}/destacadas`);
  }

  getMisResenas(): Observable<any[]> {
    const token = localStorage.getItem('token');
    return this.http.get<any[]>(`${this.baseUrl}/mis-resenas`, {
      headers: { Authorization: `Bearer ${token}` }
    });
  }

}
